tsl/src/continuous_aggs/create.c

```diff
@@ -756,6 +764,79 @@ caggtimebucket_validate(CAggTimebucketInfo *tbinfo, List *groupClause, List *tar
 				tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
 			}
 
+			if (list_length(fe->args) >= 3)
+			{
+				tz_arg = eval_const_expressions(NULL, lthird(fe->args));
+				if (!IsA(tz_arg, Const))
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("only immutable expressions allowed in time bucket function"),
+							 errhint("Use an immutable expression as third argument"
+									 " to the time bucket function.")));
+
+				Const *tz = castNode(Const, tz_arg);
+				if ((tz->consttype == TEXTOID) && (list_length(fe->args) == 3))
+				{
+					/* Timezone specified */
+					const char *tz_name = TextDatumGetCString(tz->constvalue);
+
+					if (!ts_is_valid_timezone_name(tz_name))
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("invalid timezone name \"%s\"", tz_name)));
+					}
+
+					tbinfo->timezone = tz_name;
+					tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
+				}
+				else
+				{
+					/*
+					 * Custom origin specified. This is always treated as
+					 * a variable-sized bucket case.
+					 */
+					tbinfo->bucket_width = BUCKET_WIDTH_VARIABLE;
+
+					if (tz->constisnull)
+					{
+						ereport(ERROR,
+								(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+								 errmsg("invalid origin value: null")));
+					}
+
+					switch (tz->consttype)
+					{
+						case DATEOID:
+							tbinfo->origin = DatumGetTimestamp(
+								DirectFunctionCall1(date_timestamp, tz->constvalue));
+							break;
+						case TIMESTAMPOID:
+							tbinfo->origin = DatumGetTimestamp(tz->constvalue);
+							break;
+						case TIMESTAMPTZOID:
```

Comment:
It could be the case that 3rd arg is `TIMESTAMPOID` and 4th arg is `TEXTOID`. This would be allowed by `is_valid_bucketing_function()` as long as the 2nd argument is also `TIMESTAMPOID`. But this `time_bucket_ng` overloaded function is not defined and does not exist.

Reply:
No, actually it will not be allowed. `is_valid_bucketing_function()` takes allowed functions from the func cache, and there is no function with such signature there. This is ensured by `initialize_func_info()` which places only functions listed in `FuncInfo funcinfo[]` there. Basically, is_valid_bucketing_function simply calculates a boolean value that could be stored in funcinfo[] instead.

But I agree all this may be pretty confusing. I think I will just include a refactoring we discussed previously to this PR.

Reply:
I refactored the function in the updated version of the PR 👍